File: sdk/dotnet/Ferrite.Client/VectorClient.cs
using System.Globalization;

namespace Ferrite.Client;

/// <summary>
/// Client for Ferrite vector search operations.
/// </summary>
public sealed class VectorClient
{
    private readonly FerriteClient _client;

    public VectorClient(FerriteClient client)
    {
        _client = client;
    }

    /// <summary>Creates a new vector index.</summary>
    public async Task CreateIndexAsync(string name, int dimension, string metric, CancellationToken ct = default)
    {
        await _client.ExecuteAsync(ct, "VECTOR.INDEX.CREATE", name,
            dimension.ToString(), metric).ConfigureAwait(false);
    }

    /// <summary>Adds a vector with optional metadata to an index.</summary>
    public async Task AddVectorAsync(string index, string id, float[] vector,
        Dictionary<string, object>? metadata = null, CancellationToken ct = default)
    {
        var args = new List<string> { "VECTOR.ADD", index, id, EncodeVector(vector) };
        if (metadata is { Count: > 0 })
        {
            args.Add(EncodeMetadata(metadata));
        }
        await _client.ExecuteAsync(ct, args.ToArray()).ConfigureAwait(false);
    }

    /// <summary>Searches for the nearest vectors.</summary>
    public async Task<VectorResult[]> SearchAsync(string index, float[] query, int topK,
        CancellationToken ct = default)
    {
        var val = await _client.ExecuteAsync(ct, "VECTOR.SEARCH", index,
            EncodeVector(query), topK.ToString()).ConfigureAwait(false);

        if (val is not object?[] arr)
            return Array.Empty<VectorResult>();

        var results = new List<VectorResult>();
        for (var i = 0; i + 2 < arr.Length; i += 3)
        {
            var rid = arr[i]?.ToString() ?? "";
            var distance = arr[i + 1] != null
                ? double.Parse(arr[i + 1]!.ToString()!, CultureInfo.InvariantCulture) : 0;
            Dictionary<string, object>? meta = null;
            if (arr[i + 2] is string metaStr && !string.IsNullOrEmpty(metaStr))
            {
                meta = ParseMetadata(metaStr);
            }
            results.Add(new VectorResult(rid, distance, meta));
        }
        return results.ToArray();
    }

    /// <summary>Deletes a vector from an index.</summary>
    public async Task DeleteAsync(string index, string id, CancellationToken ct = default)
    {
        await _client.ExecuteAsync(ct, "VECTOR.DELETE", index, id).ConfigureAwait(false);
    }

    private static string EncodeVector(float[] vector)
    {
        return string.Join(",", vector.Select(v => v.ToString(CultureInfo.InvariantCulture)));
    }

    private static string EncodeMetadata(Dictionary<string, object> metadata)
    {
        var pairs = metadata.Select(kv =>
        {
            var val = kv.Value is string s ? $"\"{Escape(s)}\"" : kv.Value.ToString();
            return $"\"{Escape(kv.Key)}\":{val}";
        });
        return "{" + string.Join(",", pairs) + "}";
    }

    private static Dictionary<string, object> ParseMetadata(string json)
    {
        var dict = new Dictionary<string, object>();
        var inner = json.Trim('{', '}', ' ');
        if (string.IsNullOrEmpty(inner)) return dict;

        foreach (var pair in inner.Split(','))
        {
            var parts = pair.Split(':', 2);
            if (parts.Length != 2) continue;
            var key = parts[0].Trim().Trim('"');
            var val = parts[1].Trim();
            if (val.StartsWith('"') && val.EndsWith('"'))
                dict[key] = val[1..^1];
            else if (double.TryParse(val, CultureInfo.InvariantCulture, out var num))
                dict[key] = num;
            else
                dict[key] = val;
        }
        return dict;
    }

    private static string Escape(string s) => s.Replace("\\", "\\\\").Replace("\"", "\\\"");
}

/// <summary>
/// Represents a single vector search result.
/// </summary>
public sealed record VectorResult(string Id, double Distance, Dictionary<string, object>? Metadata);

File: sdk/java/src/main/java/io/ferritelabs/ferrite/VectorClient.java
package io.ferritelabs.ferrite;

import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Client for Ferrite vector search operations.
 *
 * <pre>{@code
 * try (FerriteClient client = FerriteClient.connect("localhost", 6379)) {
 *     VectorClient vectors = new VectorClient(client);
 *     vectors.createIndex("embeddings", 384, "cosine");
 *     vectors.addVector("embeddings", "doc1", new float[]{0.1f, 0.2f}, Map.of("text", "hello"));
 *     List<VectorResult> results = vectors.search("embeddings", new float[]{0.1f, 0.2f}, 10);
 * }
 * }</pre>
 */
public class VectorClient {

    private final FerriteClient client;

    public VectorClient(FerriteClient client) {
        this.client = client;
    }

    /** Creates a new vector index with the given dimension and distance metric. */
    public void createIndex(String name, int dimension, String metric) throws IOException {
        client.executeRaw("VECTOR.INDEX.CREATE", name, String.valueOf(dimension), metric);
    }

    /** Adds a vector with optional metadata to an index. */
    public void addVector(String index, String id, float[] vector, Map<String, Object> metadata)
            throws IOException {
        List<String> args = new ArrayList<>();
        args.add("VECTOR.ADD");
        args.add(index);
        args.add(id);
        args.add(encodeVector(vector));
        if (metadata != null && !metadata.isEmpty()) {
            args.add(encodeMetadata(metadata));
        }
        client.executeRaw(args.toArray(new String[0]));
    }

    /** Searches for the nearest vectors. */
    @SuppressWarnings("unchecked")
    public List<VectorResult> search(String index, float[] query, int topK) throws IOException {
        Object val = client.executeRaw("VECTOR.SEARCH", index, encodeVector(query), String.valueOf(topK));
        if (val == null) {
            return Collections.emptyList();
        }

        List<Object> arr = (List<Object>) val;
        List<VectorResult> results = new ArrayList<>();
        for (int i = 0; i + 2 < arr.size(); i += 3) {
            String rid = arr.get(i) != null ? arr.get(i).toString() : "";
            double distance = 0;
            if (arr.get(i + 1) != null) {
                distance = Double.parseDouble(arr.get(i + 1).toString());
            }
            Map<String, Object> meta = null;
            if (arr.get(i + 2) != null) {
                meta = parseMetadata(arr.get(i + 2).toString());
            }
            results.add(new VectorResult(rid, distance, meta));
        }
        return results;
    }

    /** Deletes a vector from an index. */
    public void deleteVector(String index, String id) throws IOException {
        client.executeRaw("VECTOR.DELETE", index, id);
    }

    private static String encodeVector(float[] vector) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < vector.length; i++) {
            if (i > 0) sb.append(',');
            sb.append(vector[i]);
        }
        return sb.toString();
    }

    private static String encodeMetadata(Map<String, Object> metadata) {
        // Simple JSON encoding without external dependencies
        StringBuilder sb = new StringBuilder("{");
        boolean first = true;
        for (Map.Entry<String, Object> entry : metadata.entrySet()) {
            if (!first) sb.append(',');
            sb.append('"').append(escapeJson(entry.getKey())).append("\":");
            Object val = entry.getValue();
            if (val instanceof String) {
                sb.append('"').append(escapeJson(val.toString())).append('"');
            } else {
                sb.append(val);
            }
            first = false;
        }
        sb.append('}');
        return sb.toString();
    }

    private static Map<String, Object> parseMetadata(String json) {
        // Minimal JSON parsing for flat key-value metadata
        Map<String, Object> map = new LinkedHashMap<>();
        if (json == null || json.isEmpty() || json.equals("{}")) {
            return map;
        }
        // Strip outer braces
        String inner = json.substring(1, json.length() - 1).trim();
        if (inner.isEmpty()) return map;

        // Split by comma (simplified, doesn't handle nested objects)
        for (String pair : inner.split(",")) {
            String[] kv = pair.split(":", 2);
            if (kv.length == 2) {
                String key = kv[0].trim().replaceAll("^\"|\"$", "");
                String val = kv[1].trim();
                if (val.startsWith("\"") && val.endsWith("\"")) {
                    map.put(key, val.substring(1, val.length() - 1));
                } else {
                    try {
                        map.put(key, Double.parseDouble(val));
                    } catch (NumberFormatException e) {
                        map.put(key, val);
                    }
                }
            }
        }
        return map;
    }

    private static String escapeJson(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    /** Represents a single vector search result. */
    public static class VectorResult {
        private final String id;
        private final double distance;
        private final Map<String, Object> metadata;

        public VectorResult(String id, double distance, Map<String, Object> metadata) {
            this.id = id;
            this.distance = distance;
            this.metadata = metadata;
        }

        public String getId() { return id; }
        public double getDistance() { return distance; }
        public Map<String, Object> getMetadata() { return metadata; }

        @Override
        public String toString() {
            return "VectorResult{id='" + id + "', distance=" + distance + '}';
        }
    }
}
